src/cert/chain.rs

```diff
@@ -0,0 +1,239 @@
+/// certificate slot.
+pub trait TrustChain {
+    /// Gets the length of the `slot`th chain.
+    ///
+    /// Cannot be zero; returns `None` if this chain has no such slot.
+    fn chain_len(&self, slot: u8) -> Option<NonZeroUsize>;
+
+    /// Gets the `index`th cert of the `slot`th chain.
+    ///
+    /// Returns `None` if `index` is out of bounds or if there is no `slot`th
+    /// chain. These cases can be distinguished by calling `chain_len()`.
+    fn cert(&self, slot: u8, index: usize) -> Option<&Cert>;
+}
+
+/// A simple trust with only one slot.
+#[derive(Debug)]
+pub struct SimpleChain<'cert, const LEN: usize> {
+    chain: [Cert<'cert>; LEN],
+}
+
+impl<'cert, const LEN: usize> SimpleChain<'cert, LEN> {
+    /// Parses and verifies the trust chain described by `raw_chain`,
+    /// which starts at the root certificate for the trust chain.
+    pub fn parse(
+        raw_chain: &[&'cert [u8]],
+        format: CertFormat,
+        ciphers: &mut impl sig::Ciphers,
+    ) -> Result<Self, Error> {
+        if raw_chain.len() != LEN {
+            return Err(Error::Io(io::Error::BufferExhausted));
+        }
+
+        let mut chain = ArrayVec::new();
+        for (i, &raw_cert) in raw_chain.iter().enumerate() {
+            let prev = chain.last();
+            let key = prev.map(|cert: &Cert| cert.subject_key());
+            let cert = Cert::parse(raw_cert, format, key, ciphers)?;
+
+            let prev = prev.unwrap_or(&cert);
+            if prev.subject() != cert.issuer() {
+                return Err(Error::BadChain);
+            }
+            if !prev.supports_cert_signing() {
+                return Err(Error::BadChain);
+            }
+
+            // None is also ok; it means the format (e.g. CWT) does not support
+            // a CA bit.
+            if prev.is_ca_cert() == Some(false) {
+                return Err(Error::BadChain);
+            }
+
+            // raw_chain.len() - i is the number of certificates that follow
+            // `cert`; the path length constraint for `prev` is the number of
+            // certs that follow it, except the leaf; these numbers are the
+            // same.
+            if !prev.is_within_path_len_constraint(raw_chain.len() - i) {
+                return Err(Error::BadChain);
+            }
+
+            chain.push(cert);
+        }
+
+        // Cannot panic, since we called push() LEN times.
+        Ok(Self {
+            chain: chain.into_inner().unwrap(),
+        })
+    }
+}
+
+impl<const LEN: usize> TrustChain for SimpleChain<'_, LEN> {
+    fn cert(&self, slot: u8, index: usize) -> Option<&Cert> {
+        if slot != 0 {
+            return None;
+        }
+        self.chain.get(index)
+    }
+
+    fn chain_len(&self, slot: u8) -> Option<NonZeroUsize> {
+        if slot != 0 {
+            return None;
+        }
+
+        NonZeroUsize::new(self.chain.len())
+    }
+}
+
+#[cfg(all(test, not(miri)))]
```

Comment:
It is not at all clear to me why these are disabled when run under Miri. Is that a project-wide thing I should know? If not, please document.

Reply:
Oh, it's because of https://github.com/lowRISC/manticore/issues/103. Perhaps leave a comment here mentioning the issue?